src/platform-aws.cpp

```diff
@@ -753,51 +768,103 @@ int platform_config_endpoint(struct fi_info *info, struct fid_ep* endpoint) {
 	return ret;
 }
 
-static int get_rail_vf_idx(struct fi_info *info)
+static struct platform_aws_node_guid* get_node_guid_fields(struct fi_info *info)
 {
-	char *node_guid_filename = NULL;
-	FILE *fp = NULL;
-	int vf_idx;
-	int ret;
+	if (!info->nic || !info->nic->device_attr || !info->nic->device_attr->name) {
+		NCCL_OFI_WARN("fi_nic attributes not available.");
+		return nullptr;
+	}
 
-	ret = asprintf(&node_guid_filename, "/sys/class/infiniband/%s/node_guid",
-		       info->nic->device_attr->name);
-	if (ret < 0) {
-		vf_idx = -errno;
-		goto cleanup;
+	std::string device_name = info->nic->device_attr->name;
+
+	/* Check to see if we've already parsed the fields for this RDMA device */
+	{
+		std::lock_guard<std::mutex> lock(cache_mutex);
+		auto it = guid_cache.find(device_name);
+		if (it != guid_cache.end()) {
+			return &(it->second);
+		}
 	}
-	fp = fopen(node_guid_filename, "r");
-	if (fp == NULL) {
-		NCCL_OFI_WARN("Error opening file: %s", node_guid_filename);
-		vf_idx = -errno;
-		goto cleanup;
+
+	std::string filepath = "/sys/class/infiniband/";
+	filepath += info->nic->device_attr->name;
+	filepath += "/node_guid";
+
+	std::ifstream file(filepath);
+	if (!file.is_open()) {
+		throw std::system_error(errno, std::system_category(),
+			"Failed to open " + filepath);
 	}
 
-	/**
-	 * GUID is a 64-bit hex number with format:
-	 *
-	 * XXXX:XXXX:XXXX:XXXX
-	 *
-	 * The lowest 8 bits are the VF id.
-	 */
-	ret = fscanf(fp, "%*x:%*x:%*x:%*2x%2x", &vf_idx);
-	if (ret != 1) {
-		NCCL_OFI_WARN("GUID parsing failed, got %d, expected 1", ret);
-		vf_idx = -EIO;
-		goto cleanup;
+	std::string guid_str;
+	if (!std::getline(file, guid_str)) {
+		throw std::runtime_error("Failed to read data from " + filepath);
 	}
 
-cleanup:
-	if (node_guid_filename) {
-		free(node_guid_filename);
+	/* Parse the GUID string in XXXX:XXXX:XXXX:XXXX format */
+	unsigned int a, b, c, d;
+	if (sscanf(guid_str.c_str(), "%4x:%4x:%4x:%4x", &a, &b, &c, &d) != 4) {
+		throw std::runtime_error("Invalid GUID format in " + filepath);
 	}
-	if (fp != NULL) {
-		fclose(fp);
+
+	/* Reconstruct the 64-bit value */
+	uint64_t raw_value = ((uint64_t)a << 48) |
+			    ((uint64_t)b << 32) |
+			    ((uint64_t)c << 16) |
+			    ((uint64_t)d);
+
+	NCCL_OFI_INFO(NCCL_INIT, "GUID of %s: %016lx", info->nic->device_attr->name, raw_value);
+
+	/*
+	 * +---------------------+------------------+------------+------------+
+	 * |63                 24|23              16|15         8|7          0|
+	 * +---------------------+------------------+------------+------------+
+	 * | func_mac high bytes | per-card pci_bus | vf pci_bus |  func_idx  |
+	 * +---------------------+------------------+------------+------------+
```

Comment:
Gah, need to fix comment, but will do that along with any other changes to address review comments.